Simplenote/AuthenticationValidator.swift

```diff
@@ -0,0 +1,132 @@
+import Foundation
+
+
+// MARK: - AuthenticationValidator
+//
+struct AuthenticationValidator {
+
+    /// We're enhancing password strength requirements. But legacy accounts must be allowed in. That's why we have two sets of rules!
+    ///
+    let style: Style
+
+    /// Minimum Password Length: Login
+    ///
+    private let loginPasswordLength  = UInt(4)
+
+    /// Minimum Password Length: SignUp
+    ///
+    private let signupPasswordLength = UInt(8)
+
+    /// Defines the minimum allowed password length
+    ///
+    private var minimumPasswordLength: UInt {
+        return (style == .login) ? loginPasswordLength : signupPasswordLength
+    }
+
+
+    /// Returns the Validation Result for a given Username
+    ///
+    func performUsernameValidation(username: String) -> Result {
+        let predicate = NSPredicate.predicateForEmailValidation()
+        return predicate.evaluate(with: username) ? .success : .emailInvalid
+    }
+
+    /// Returns the Validation Result for a given Password (with its associated Username)
+    ///
+    func performPasswordValidation(username: String, password: String) -> Result {
+        guard password.count >= minimumPasswordLength else {
+            return .passwordTooShort(length: minimumPasswordLength)
+        }
+
+        guard style == .signup else {
+            return .success
+        }
+
+        guard password != username else {
+            return .passwordMatchesUsername
+        }
+
+        guard !password.contains(String.newline), !password.contains(String.tab) else {
+            return .passwordContainsInvalidCharacter
+        }
+
+        return .success
+    }
+}
+
+
+// MARK: - Nested Types
+//
+extension AuthenticationValidator {
+
+    enum Style {
+        case login
+        case signup
+    }
```

Comment:
Renamed to `Style` > `strong` / `legacy` (since, now that I think about it!!, that enum shouldn't know / care about who uses it, or where).

Thank you!!!